File: crates/crm/src/domain/comment.rs
//! Domain (and wire) models for CRM comment threads.
//!
//! A CRM comment thread mirrors the document `Thread`/`Comment` shape closely
//! enough that the frontend reuses its thread-assembly and rendering logic,
//! but threads hang off a CRM company or contact (not a document) and use
//! uuid ids. `entityType` + `entityId` replace the document `documentId`.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;

/// Which CRM entity a comment thread is attached to. Serializes to
/// `crm_company` / `crm_contact` — matching the `entityType` the frontend
/// uses elsewhere when building entity URLs — and is parsed from the
/// `{entity_type}` path segment on the comment routes.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "axum", derive(utoipa::ToSchema))]
#[serde(rename_all = "snake_case")]
pub enum CrmCommentEntityType {
    /// Thread attached to a `crm_companies` row.
    CrmCompany,
    /// Thread attached to a `crm_contacts` row.
    CrmContact,
}

/// A CRM comment thread: the parent record one or more comments hang off.
#[derive(Debug, Clone, Serialize)]
#[cfg_attr(feature = "axum", derive(utoipa::ToSchema))]
#[serde(rename_all = "camelCase")]
pub struct CrmThread {
    /// The thread id.
    pub thread_id: Uuid,
    /// Which CRM entity kind this thread belongs to.
    pub entity_type: CrmCommentEntityType,
    /// The id of the CRM company or contact this thread belongs to.
    pub entity_id: Uuid,
    /// Macro user id of the thread creator.
    pub owner: String,
    /// Whether the thread is resolved.
    pub resolved: bool,
    /// Arbitrary client metadata.
    pub metadata: Option<Value>,
    /// When the thread was created.
    pub created_at: DateTime<Utc>,
    /// When the thread was last updated.
    pub updated_at: DateTime<Utc>,
    /// When the thread was soft-deleted, if ever.
    pub deleted_at: Option<DateTime<Utc>>,
}

/// A single comment within a [`CrmThread`].
#[derive(Debug, Clone, Serialize)]
#[cfg_attr(feature = "axum", derive(utoipa::ToSchema))]
#[serde(rename_all = "camelCase")]
pub struct CrmComment {
    /// The comment id.
    pub comment_id: Uuid,
    /// The id of the thread this comment belongs to.
    pub thread_id: Uuid,
    /// Optional explicit ordering within the thread; the frontend falls
    /// back to `createdAt` when absent.
    pub order: Option<i32>,
    /// Macro user id of the comment author.
    pub owner: String,
    /// Macro user id of the actual sender, when distinct from `owner`.
    pub sender: Option<String>,
    /// The comment body (markdown).
    pub text: String,
    /// Arbitrary client metadata.
    pub metadata: Option<Value>,
    /// When the comment was created.
    pub created_at: DateTime<Utc>,
    /// When the comment was last updated.
    pub updated_at: DateTime<Utc>,
    /// When the comment was soft-deleted, if ever.
    pub deleted_at: Option<DateTime<Utc>>,
}

/// A [`CrmThread`] with its comments nested under it — the unit the
/// frontend renders.
#[derive(Debug, Clone, Serialize)]
#[cfg_attr(feature = "axum", derive(utoipa::ToSchema))]
#[serde(rename_all = "camelCase")]
pub struct CrmCommentThread {
    /// The thread.
    pub thread: CrmThread,
    /// The thread's comments, oldest first.
    pub comments: Vec<CrmComment>,
}

/// Outcome of soft-deleting a CRM comment: reports whether the parent thread
/// was soft-deleted too (it is when the deleted comment was its last live one).
#[derive(Debug, Clone, Serialize)]
#[cfg_attr(feature = "axum", derive(utoipa::ToSchema))]
#[serde(rename_all = "camelCase")]
pub struct DeleteCrmCommentResult {
    /// The deleted comment's id.
    pub comment_id: Uuid,
    /// The thread the comment belonged to.
    pub thread_id: Uuid,
    /// Whether the thread itself was soft-deleted because no live comments
    /// remained.
    pub thread_deleted: bool,
}

File: packages/sdk/src/entities/crm/comment.ts
import type {
  CrmComment as CrmCommentRecord,
  CrmThread,
  DeleteCrmCommentResult,
} from '../../../generated/storage/types.gen';
import { unwrap } from '../../utils';
import type { MacroClient } from '../../utils/client';

/** A CRM comment thread with its comments, as returned for a company or contact. */
export interface CrmThreadWithComments {
  /** The thread. */
  thread: CrmThread;
  /** The thread's comments, oldest first. */
  comments: CrmComment[];
}

/**
 * A comment on a CRM company or contact. Shared by both entity types, since
 * the underlying API and record shape are already unified across them.
 */
export class CrmComment {
  private constructor(
    private readonly client: MacroClient,
    readonly id: string,
    private record: CrmCommentRecord,
  ) {}

  /** Wrap a record already in hand (e.g. from a list or create response). */
  static from(client: MacroClient, record: CrmCommentRecord): CrmComment {
    return new CrmComment(client, record.commentId, record);
  }

  /** The comment body (markdown). */
  get text(): string {
    return this.record.text;
  }

  /** The id of the thread this comment belongs to. */
  get threadId(): string {
    return this.record.threadId;
  }

  /** Macro user id of the comment author. */
  get owner(): string {
    return this.record.owner;
  }

  /** Macro user id of the actual sender, when distinct from {@link owner}. */
  get sender(): string | undefined {
    return this.record.sender ?? undefined;
  }

  /** Explicit ordering within the thread, if set. */
  get order(): number | undefined {
    return this.record.order ?? undefined;
  }

  /** Arbitrary client metadata attached to the comment. */
  get metadata(): unknown {
    return this.record.metadata;
  }

  /** When the comment was created. */
  get createdAt(): string {
    return this.record.createdAt;
  }

  /** When the comment was last updated. */
  get updatedAt(): string {
    return this.record.updatedAt;
  }

  /** When the comment was soft-deleted, if ever. */
  get deletedAt(): string | undefined {
    return this.record.deletedAt ?? undefined;
  }

  /** Replace this comment's text (markdown). */
  async edit(text: string): Promise<this> {
    this.record = unwrap(
      await this.client.storage.editCrmComment({
        path: { comment_id: this.id },
        body: { text },
      }),
    );
    return this;
  }

  /** Soft-delete this comment; the thread goes too when it was the last live one. */
  async delete(): Promise<DeleteCrmCommentResult> {
    return unwrap(
      await this.client.storage.deleteCrmComment({
        path: { comment_id: this.id },
      }),
    );
  }
}
